src/fhirtypes/ElementDefinition.ts

```diff
@@ -414,6 +416,101 @@ export class ElementDefinition {
+    // The constrainType() method applies a type constraint to an existing
+    // ElementDefinition.type. Since this is a new ElementDefinition, it
+    // does not yet have a 'type', so we need to assign an "initial" value
+    // that the constrainType() method can process.
+    this.type = this.initializeElementType(rule, fisher);
+    const target = this.structDef.getReferenceName(rule.path, this);
+    this.constrainType(rule, fisher, target);
+
+    this.constrainCardinality(rule.min, rule.max);
+
+    this.applyFlags(
+      rule.mustSupport,
+      rule.summary,
+      rule.modifier,
+      rule.trialUse,
+      rule.normative,
+      rule.draft
+    );
+
+    // Since 'short' and 'definition' are plain string values, directly assign the values.
+    if (!isEmpty(rule.short)) {
+      this.short = rule.short;
+    }
+    if (isEmpty(rule.definition)) {
+      if (!isEmpty(rule.short)) {
+        // According to sdf-3 in http://hl7.org/fhir/structuredefinition.html#invs,
+        // definition should be provided. Default it here if possible.
+        this.definition = rule.short;
+      }
+    } else {
+      this.definition = rule.definition;
+    }
+
+    // Add the base attribute
+    this.base = {
+      path: `${this.structDef.type}.${rule.path}`,
+      min: rule.min,
+      max: rule.max
+    };
+
+    // Add the 'ele-1' constraint attribute
+    this.constraint = [
+      {
+        key: 'ele-1',
+        severity: 'error',
+        human: 'All FHIR elements must have a @value or children',
+        expression: 'hasValue() or (children().count() > id.count())',
+        xpath: '@value|f:*|h:div',
+        source: 'http://hl7.org/fhir/StructureDefinition/Element'
+      }
+    ];
+  }
+
+  /**
+   * Define and return the initial ElementDefinition.type from the AddElementRule.
+   * @param {AddElementRule} rule - specific instance of the rule
+   * @param {Fishable} fisher - A fishable implementation for finding definitions and metadata
+   * @returns {ElementDefinitionType[]} The element types as defined by the AddElementRule
+   * @private
+   */
+  private initializeElementType(rule: AddElementRule, fisher: Fishable): ElementDefinitionType[] {
+    const initialTypes: ElementDefinitionType[] = [];
+    // If there are multiple reference types, only one EDT is created having
+    // one or more targetProfiles for the single reference type
+    let referenceEdt: ElementDefinitionType;
+    const targetProfiles: string[] = [];
+    rule.types.forEach(t => {
+      if (t.isReference) {
+        referenceEdt = referenceEdt ?? new ElementDefinitionType('Reference');
+        const metadata = fisher.fishForMetadata(t.type, Type.Resource, Type.Profile);
+        if (metadata) {
+          targetProfiles.push(metadata.url);
+        } else {
+          throw new TypeNotFoundError(t.type);
+        }
+      } else {
+        const edt = new ElementDefinitionType(t.type);
+        initialTypes.push(edt);
```

Comment:
What if the type provided by the rule is actually a _profile_?  For example, I *think* this is valid?
```
* patient 1..1 USCorePatientProfile "A US Core Patient"
```
I know this would *typically* be a reference, but in a logical model, anything goes.

Reply:
```* patient 1..1 USCorePatientProfile "A US Core Patient"```
error The type "USCorePatientProfile" does not match any of the allowed types: USCorePatientProfile

```* patient 1..1 http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient "A US Core Patient"```
error The type "http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient" does not match any of the allowed types: http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient

```* patient 1..1 us-core-patient "A US Core Patient"```
SUCCESS ->
```
      {
        "id": "CustomLogicalModel.patient",
        "path": "CustomLogicalModel.patient",
        "short": "A US Core Patient",
        "definition": "A US Core Patient",
        "min": 1,
        "max": "1",
        "base": {
          "path": "CustomLogicalModel.patient",
          "min": 1,
          "max": "1"
        },
        "type": [
          {
            "code": "Patient",
            "profile": [
              "http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient"
            ]
          }
        ],
        "constraint": [
          {
            "key": "ele-1",
            "severity": "error",
            "human": "All FHIR elements must have a @value or children",
            "expression": "hasValue() or (children().count() > id.count())",
            "xpath": "@value|f:*|h:div",
            "source": "http://hl7.org/fhir/StructureDefinition/Element"
          }
        ]
      },
```
ElementDefinition.findTypeMatch() only does matching logic based on the lineage.id (not name or url).

@cmoesel Not sure what you might want to do with this. Maybe a Jira issue on the backlog??

Reply:
I think the issue is that our code expects `type.code` to be an `id` of a FHIR type.  In most cases (at least up until logical models) this was guaranteed to be true.  Consider the definition of `type.code`:
> URL of Data type or Resource that is a(or the) type used for this element. References are URLs that are relative to http://hl7.org/fhir/StructureDefinition e.g. "string" is a reference to http://hl7.org/fhir/StructureDefinition/string. Absolute URLs are only allowed in logical models.

I expect this is why using the name or url did not work -- because it was getting put into `type.code` and then the `constrainType` function could not look it up (since it thinks it should be an `id`).  I think maybe this would be fixed if you looked up the type (by whatever was passed in) to determine if it is a specialization or a constraint.  If it's a constraint, then the `type.code` should be the `SD.type` and the `type.profile` array should contain the `SD.url`.

Reply:
I updated my last comment in light of the approach I am suggesting in the next comment.